src/main/groovy/org/spongepowered/asm/gradle/plugins/MixinEclipse.groovy

```diff
@@ -0,0 +1,170 @@
+            order.add(key)
+            return super.put(key, value)
+        }
+        
+        public Object arg(String key, String value) {
+            return put('org.eclipse.jdt.apt.processorOptions/' + key, value)
+        }
+    }
+    
+    static class EclipseJdtAptTask extends DefaultTask {
+        @InputFile File mappingsIn
+        @Input File refmapOut = project.file("build/${name}/mixins.refmap.json")
+        @Input File mappingsOut = project.file("build/${name}/mixins.mappings.tsrg")
+        @Input Map<String, String> processorOptions = new TreeMap<>()
+        
+        @Input File genTestDir = project.file('build/.apt_generated_test')
+        @Input File genDir = project.file('build/.apt_generated')
+        
+        @OutputFile File output
+        
+        
+        @TaskAction
+        def run() {
+            MixinExtension extension = project.extensions.findByType(MixinExtension.class)
+            def props = new OrderedProperties()
+            props.put('eclipse.preferences.version', '1')
+            props.put('org.eclipse.jdt.apt.aptEnabled', 'true')
+            props.put('org.eclipse.jdt.apt.reconcileEnabled', 'true')
+            props.put('org.eclipse.jdt.apt.genSrcDir', genDir.canonicalPath)
+            props.put('org.eclipse.jdt.apt.genSrcTestDir', genTestDir.canonicalPath)
+            props.arg('reobfTsrgFile', mappingsIn.canonicalPath)
+            props.arg('outTsrgFile', mappingsOut.canonicalPath)
+            props.arg('outRefMapFile', refmapOut.canonicalPath)
+            props.arg('pluginVersion', MixinGradlePlugin.VERSION)
+            
+            if (extension.disableTargetValidator) {
+                props.arg('disableTargetValidator', 'true')
+            }
+            
+            if (extension.disableTargetExport) {
+                props.arg('disableTargetExport', 'true')
+            }
+            
+            if (extension.disableOverwriteChecker) {
+                props.arg('disableOverwriteChecker', 'true')
+            }
+            
+            if (extension.overwriteErrorLevel != null) {
+                props.arg('overwriteErrorLevel', extension.overwriteErrorLevel.toString().trim())
+            }
+            
+            if (extension.defaultObfuscationEnv != null) {
+                props.arg('defaultObfuscationEnv', extension.defaultObfuscationEnv)
+            }
+            
+            if (extension.mappingTypes.size() > 0) {
+                props.arg('mappingTypes', extension.mappingTypes.join(','))
+            }
+
+            if (extension.tokens.size() > 0) {
+                props.arg('tokens', extension.tokens.collect { token -> token.key + ' ' + token.value }.join(';'))
+            }
+            
+            if (extension.extraMappings.size() > 0) {
+                props.arg('reobfTsrgFiles', extension.extraMappings.collect { file -> project.file(file).toString() }.join(';'))
+            }
+
+            /* TODO... Not sure what this is
```

Comment:
"Imports" hacks around an issue with multi-project setups where the AP doesn't have info from upstream projects when running on a downstream one, or using mixins from a dependency which contains mixins with derived types in the current project. For example if a dependency adds `MixinEntity` and this project contains `MixinSomeDerivedEntity extends MixinEntity` from the other project, this ensures that the AP knows about the other mixins. `generateImportsFile` uses the logic in [`Imports`](https://github.com/SpongePowered/MixinGradle/blob/master/src/main/groovy/org/spongepowered/asm/gradle/plugins/meta/Imports.groovy) to scan jars for mixin targets.

This will probably go away in the future because it's becoming increasingly clear that incremental compilation (and the corresponding (now invalid) assumption that all mixin classes are included in the AP pass) means that the AP is probably going to have to be refactored to generate some internal database with information generated solely via Mirror.